lib/resolve-imports-exports.js

```diff
@@ -0,0 +1,208 @@
+var path = require('path');
+var startsWith = require('string.prototype.startswith');
+
+function validateExports(exports, basePath) {
+    var isConditional = true;
+
+    if (typeof exports === 'object' && !Array.isArray(exports)) {
+        var exportKeys = Object.keys(exports);
+
+        for (var i = 0; i < exportKeys.length; i++) {
+            var isKeyConditional = exportKeys[i][0] !== '.';
+            if (i === 0) {
+                isConditional = isKeyConditional;
+            } else if (isKeyConditional !== isConditional) {
+                var err = new Error('Invalid package config ' + basePath + path.sep + 'package.json, '
+                    + '"exports" cannot contain some keys starting with \'.\' and some not. '
+                    + 'The exports object must either be an object of package subpath keys '
+                    + 'or an object of main entry condition name keys only.');
+                err.code = 'ERR_INVALID_PACKAGE_CONFIG';
+                throw err;
+            }
+        }
+    }
+
+    if (isConditional) {
+        return { '.': exports };
+    } else {
+        return exports;
+    }
+}
+
+function validateConditions(names) {
+    // If exports contains any index property keys, as defined in ECMA-262 6.1.7 Array Index, throw an Invalid Package Configuration error.
+
+    for (var i = 0; i < names.length; i++) {
+        var name = names[i];
+        var nameNum = Number(name);
+
+        if (String(nameNum) === name && nameNum >= 0 && nameNum < 0xFFFFFFFF) {
+            var err = new Error('"exports" cannot contain numeric property keys');
+            err.code = 'ERR_INVALID_PACKAGE_CONFIG';
+            throw err;
+        }
+    }
+
+    return names;
+}
+
+function resolvePackageTarget(packagePath, parent, key, target, subpath, internal, conditions) {
+    if (typeof target === 'string') {
+        var resolvedTarget = path.resolve(packagePath, target);
+        var invalidTarget = false;
+
+        if (!startsWith(target, './')) {
+            if (!internal) {
+                invalidTarget = true;
+            } else if (!startsWith(target, '../') && !startsWith(target, '/')) {
+                invalidTarget = true;
+            } else {
+                // TODO: imports need call package_resolve here
+            }
+        }
+
+        var targetParts = target.split(/[\\/]/).slice(1); // slice to strip the leading '.'
+        if (invalidTarget || targetParts.indexOf('node_modules') !== -1 || targetParts.indexOf('.') !== -1 || targetParts.indexOf('..') !== -1) {
```

Comment:
hmm, i'm a bit confused. if i have a RHS that's `./foo/../a.js`, that should be valid - or at this point is `target` path-normalized?

Reply:
A RHS like that is invalid cf. nodejs docs

> If target split on "/" or "\\" contains any ".", ".." or "node_modules" segments after the first segment, throw an Invalid Package Target error.
>
> ~ PACKAGE_TARGET_RESOLVE step 1.3 https://nodejs.org/dist/latest-v14.x/docs/api/esm.html#esm_resolver_algorithm

Reply:
ah awesome, i probably was the one to ask for that in the first place (－‸ლ)